src/libdredd/src/mutation_replace_expr.cc

```diff
@@ -0,0 +1,289 @@
+      mutant_offset++;
+    }
+
+    if (exprType.isFloatingPoint()) {
+      // Replace floating point expression with 0.0
+      new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                   << mutant_offset << ")) return 0.0;\n";
+      mutant_offset++;
+
+      // Replace floating point expression with 1.0
+      new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                   << mutant_offset << ")) return 1.0;\n";
+      mutant_offset++;
+
+      // Replace floating point expression with -1.0
+      new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                   << mutant_offset << ")) return -1.0;\n";
+      mutant_offset++;
+    }
+  }
+}
+
+std::string MutationReplaceExpr::GenerateMutatorFunction(
+    clang::ASTContext& ast_context, const std::string& function_name,
+    const std::string& result_type, const std::string& input_type,
+    int& mutation_id) const {
+  std::stringstream new_function;
+  new_function << "static " << result_type << " " << function_name << "(";
+  if (ast_context.getLangOpts().CPlusPlus) {
+    new_function << "std::function<" << input_type << "()>";
+  } else {
+    new_function << input_type;
+  }
+  new_function << " arg, int local_mutation_id) {\n";
+
+  std::string arg_evaluated = "arg";
+  if (ast_context.getLangOpts().CPlusPlus) {
+    arg_evaluated += "()";
+  } else {
+    if (expr_.isLValue()) {
+      //      arg_evaluated = "(*" + arg_evaluated + ")";
+    }
+  }
+
+  // Quickly apply the original operator if no mutant is enabled (which will be
+  // the common case).
+  new_function << "  if (!__dredd_some_mutation_enabled) return "
+               << arg_evaluated << ";\n";
+
+  int mutant_offset = 0;
+  const clang::BuiltinType* exprType =
+      expr_.getType()->getAs<clang::BuiltinType>();
+
+  GenerateUnaryOperatorInsertion(arg_evaluated, expr_, *exprType, new_function,
+                                 mutant_offset);
+  GenerateConstantReplacement(expr_, *exprType, ast_context, new_function,
+                              mutant_offset);
+
+  new_function << "  return " << arg_evaluated << ";\n";
+  new_function << "}\n\n";
+
+  mutation_id += mutant_offset;
+
+  return new_function.str();
+}
+
+void MutationReplaceExpr::ApplyCppTypeModifiers(const clang::Expr* expr,
+                                                std::string& type) {
+  if (expr->isLValue()) {
+    type += "&";
+    clang::QualType qualified_type = expr->getType();
+    if (qualified_type.isVolatileQualified()) {
+      type = "volatile " + type;
+    }
+  }
+}
+
+void MutationReplaceExpr::ApplyCTypeModifiers(const clang::Expr* expr,
+                                              std::string& type) {
+  if (expr->isLValue()) {
+    type += "*";
+    clang::QualType qualified_type = expr->getType();
+    if (qualified_type.isVolatileQualified()) {
+      type = "volatile " + type;
+    }
+  }
+}
```

Comment:
It looks like these are identical to the same functions in replace unary (though I didn't check for sure).

If they are, then make them public static methods of MutationReplaceExpr, and have MutationReplaceUnaryExpr call them via MutationReplaceExpr::...

Reply:
Refactored.